Interactive API Experiment - Pytorch Re-ID on Market

katerina-merkulova opened the pull request:
Tutorial experiment to Interactive API - Pytorch Re-ID (Resnet50 and Arcface loss) on Market1501 dataset.

katerina-merkulova committed:
add re-id experiment with pytorch

katerina-merkulova committed:
add re-id experiment with pytorch

katerina-merkulova committed:
add re-id experiment with pytorch

katerina-merkulova committed:
add re-id experiment with pytorch

openfl-helper commented:
Can one of the admins verify this patch?

katerina-merkulova commented:
I have read the CLA Document and I hereby sign the CLA

alexey-gruzdev commented:
ok to test

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/director_folder/director_config.yaml:
```diff
@@ -0,0 +1,4 @@
+settings:
+  listen_ip: nnlicv838.inn.intel.com
```
Please remove all private information regarding your local setup. This address can probably be replaced with localhost.

katerina-merkulova replied:
fixed

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/one/market_shard_descriptor.py:
```diff
@@ -0,0 +1,142 @@
+
+    def __getitem__(self, index: int):
+        """Return a item by the index."""
+        imgs_path = list(self.train_dir.glob('*.jpg'))[self.rank_worldsize[0] - 1::self.rank_worldsize[1]]
```
Consider introducing a class field for this sliced list in order to avoid repeated computation of `glob`  for each item. Also, the same list is used in the `__len__` function, so the field could be reused.

katerina-merkulova replied:
thanks, fixed

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/one/start_envoy.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+set -e
+
+fx envoy start -n env_one --disable-tls -d nnlicv838.inn.intel.com:50051
```
Please remove all private information regarding your local setup. This address can probably be replaced with localhost.

itrushkin replied:
```
Usage: fx envoy start [OPTIONS]
Try 'fx envoy start --help' for help.

Error: No such option: -d
```

itrushkin replied:
In latest `develop` branch version we have introduced 2 new parameters: `-dh` (director host) and `-dp` (director port). They have replaced `-d`.

katerina-merkulova replied:
got it, fixed, thank you

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/two/market_shard_descriptor.py:
```diff
@@ -0,0 +1,140 @@
+
+    def __getitem__(self, index: int):
+        """Return a item by the index."""
+        imgs_path = list(self.train_dir.glob('*.jpg'))[self.rank_worldsize[0] - 1::self.rank_worldsize[1]]
```
Consider introducing a class field for this sliced list in order to avoid repeated computation of glob for each item. Also, the same list is used in the __len__ function, so the field could be reused.

katerina-merkulova replied:
fixed as well

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/two/start_envoy.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+set -e
+
+fx envoy start -n env_two --disable-tls -d nnlicv838.inn.intel.com:50051
```
Please remove all private information regarding your local setup. This address can probably be replaced with localhost.

katerina-merkulova replied:
done

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/workspace/transforms.py:
```diff
@@ -0,0 +1,91 @@
+from torchvision.transforms import Compose, RandomHorizontalFlip, ToTensor, Normalize, Resize
+
+
+class RandomCroping(object):
```
This transform name doesn't reflect that images get resized before cropping.

katerina-merkulova replied:
renamed to ResizeRandomCropping

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/workspace/transforms.py:
```diff
@@ -0,0 +1,91 @@
+        y_maxrange = new_height - self.height
+        x1 = int(round(random.uniform(0, x_maxrange)))
+        y1 = int(round(random.uniform(0, y_maxrange)))
+        croped_img = resized_img.crop((x1, y1, x1 + self.width, y1 + self.height))
```
```suggestion
        cropped_img = resized_img.crop((x1, y1, x1 + self.width, y1 + self.height))
```

katerina-merkulova replied:
thank you, fixed as well

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/workspace/transforms.py:
```diff
@@ -0,0 +1,91 @@
+        y1 = int(round(random.uniform(0, y_maxrange)))
+        croped_img = resized_img.crop((x1, y1, x1 + self.width, y1 + self.height))
+
+        return croped_img
```
```suggestion
        return cropped_img
```

katerina-merkulova replied:
thank you, fixed as well

dmitryagapov committed:
Create global variables for CA files/paths (#150)

* add variable for ca files

* rename variable

* divide openfl cert from ca cert folders

katerina-merkulova committed:
fix lint

katerina-merkulova committed:
fix lint

katerina-merkulova committed:
fix lint

katerina-merkulova committed:
fix lint

igor-davidyuk commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/two/market_shard_descriptor.py:
```diff
@@ -0,0 +1,153 @@
+# Copyright (C) 2020-2021 Intel Corporation
```
There is no need to keep two identical Shard Descriptors. I suggest removing one/ and two/ folders under `envoy_folder` directory and keep just two different configs files

katerina-merkulova replied:
ok, done

itrushkin commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/one/market_shard_descriptor.py:
```diff
@@ -0,0 +1,142 @@
+from openfl.interface.interactive_api.shard_descriptor import ShardDescriptor
+
+
+DATAPATH = list(Path.cwd().parents[2].rglob('**/Market'))[0]    # parent directory of project
```
`IndexError: list index out of range`

katerina-merkulova replied:
added comment below to download data and put to directory previously

itrushkin replied:
Can data be downloaded automatically?

katerina-merkulova replied:
No, I found two sources: google-drive and kaggle, but both of them are pretty hard to download automatically

github-actions[bot] committed:
@alexey-gruzdev has signed the CLA from Pull Request #156

katerina-merkulova committed:
fixes after review

katerina-merkulova committed:
fixes after review

aleksandr-mokrov committed:
Full traceback call stack on broad exception handling. (#157)

katerina-merkulova committed:
fix interface

katerina-merkulova committed:
fix interface

katerina-merkulova committed:
fix shard decriptor logic

katerina-merkulova committed:
fix shard decriptor logic

katerina-merkulova committed:
minor fixes

katerina-merkulova committed:
minor fixes

igor-davidyuk commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/market_shard_descriptor.py:
```diff
@@ -0,0 +1,139 @@
+# URL: https://www.kaggle.com/pengcw1/market-1501
+
+# search in whole project directory
+DATAPATH = list(Path.cwd().parents[2].rglob('**/Market'))[0]    # parent directory of project
```
Parametrize `DATAPATH `

katerina-merkulova replied:
ok

katerina-merkulova replied:
done

igor-davidyuk commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/market_shard_descriptor.py:
```diff
@@ -0,0 +1,139 @@
+        self.train_dir = self.dataset_dir / 'bounding_box_train'
+        self.query_dir = self.dataset_dir / 'query'
+        self.gal_dir = self.dataset_dir / 'bounding_box_test'
+        self.imgs_path = list(self.train_dir.glob('*.jpg'))[self.rank - 1::self.worldsize]
```
seem like you do not need this field, you use `self.train` anyway

katerina-merkulova replied:
I use it in len and get_item functions

katerina-merkulova replied:
fixed

igor-davidyuk commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/market_shard_descriptor.py:
```diff
@@ -0,0 +1,139 @@
+
+        self._check_before_run()
+
+        self.train, self.num_train_pids, self.num_train_imgs = self._process_dir(self.train_dir)
```
`train` list actually contains local paths to data samples which is not safe.
We can leave `query` and `gallery` as is, but try to fill train data loader using `Shard Descriptor's` __getitem__

katerina-merkulova replied:
fixed, check please

katerina-merkulova committed:
fix getitem logic

katerina-merkulova committed:
fix getitem logic

katerina-merkulova committed:
fix lint

katerina-merkulova committed:
fix lint

igor-davidyuk commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/workspace/Market_with_Director.ipynb:
```diff
@@ -0,0 +1,2183 @@
+    "        else:\n",
+    "            batch_size = 512\n",
+    "        \n",
+    "        self.shard_descriptor.set_mode('query')\n",
```
We probably should pass `shard_descriptor` to DataLoader by deep copying it. 
When the Pytorch DataLoader will try to access shard_descriptor the mode may be already changed

katerina-merkulova replied:
added shard descriptor's copies to gallery and query loaders

igor-davidyuk commented:
Could you also clear all cell outputs

katerina-merkulova committed:
rm cell outputs

katerina-merkulova committed:
rm cell outputs

aleksandr-mokrov committed:
Configure envoy health check period form director (#153)

* Configure envoy health check period form director

* Period in seconds comment

* Rename listen_ip -> listen_host. Add params to default director.yaml

* Refactor health-check

katerina-merkulova committed:
add shard descriptor's copy

katerina-merkulova committed:
add shard descriptor's copy

katerina-merkulova committed:
minor changes

katerina-merkulova committed:
minor changes

igor-davidyuk commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/start_envoy.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+set -e
+
+fx envoy start -n env_one --disable-tls -dh localhost -dp 50051
```
```suggestion
fx envoy start -n env_one --disable-tls -dh localhost -dp 50051  -sc shard_config_one.yaml
```

With this key, it is obvious how to change the `Shard Descriptor` config.

katerina-merkulova committed:
add downloading

katerina-merkulova committed:
add downloading

katerina-merkulova committed:
add shard_config key to start_envoy

Co-authored-by: igor-davidyuk <76463150+igor-davidyuk@users.noreply.github.com>

katerina-merkulova committed:
add shard_config key to start_envoy

Co-authored-by: igor-davidyuk <76463150+igor-davidyuk@users.noreply.github.com>

katerina-merkulova committed:
fix lint

katerina-merkulova committed:
fix lint

katerina-merkulova committed:
fix lint

katerina-merkulova committed:
fix lint

igor-davidyuk commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/market_shard_descriptor.py:
```diff
@@ -0,0 +1,117 @@
+    def download():
+        """Download Market1501 dataset."""
+        url = 'https://drive.google.com/uc?id=0B8-rUzbwVRk0c054eEozWG9COHM'
+        output = 'Market.zip'
```
Please add a check if the file already exists

katerina-merkulova replied:
done

igor-davidyuk commented on openfl-tutorials/interactive_api/Pytorch_Market_Re-ID/envoy_folder/market_shard_descriptor.py:
```diff
@@ -0,0 +1,117 @@
+        output = 'Market.zip'
+        gdown.download(url, output, quiet=False)
+
+        with zipfile.ZipFile(output, 'r') as zip_ref:
```
Please add a check if the folder already exists

katerina-merkulova replied:
done

katerina-merkulova committed:
add check if dataset folder exists

katerina-merkulova committed:
add check if dataset folder exists

katerina-merkulova committed:
improve check if dataset folder exists

katerina-merkulova committed:
improve check if dataset folder exists

itrushkin committed:
Tests for interactive API (#151)

* Complete Tensorflow MNIST test

* Complete PyTorch UNet Kvasir test

* Add run script for tensorflow_mnist experiment

* Add run script for pytorch_kvasir_unet experiment

* Update requirements for pytorch_kvasir_unet experiment

* Add -r key

* Add scikit-image to shard descriptor requirements

* Convert enforce_image_hw to int

* Update envoy bash script

* Pass director port as number

* Move imports to file header

* Remove unused kwargs parameter

* Add train_bs and valid_bs properties

* Split rank_worldsize variable

* Update file extension filtering

* Split rank_worldsize field

* Latest develop changes

* Remove backslash

* Fix argument order

* Remove unused kwargs

* Remove certificate-related parameters in non-TLS environment

itrushkin committed:
Unbalanced dataset splits (#125)

* Add new log level - METRIC. Write metrics on aggregator to tensorboard

* Refactor code. Add tensorboardX to setup.py

* fix flake8

* Return back global logger in native.py

* Delete raising exceptions in addLoggingLevel function

* Modify log level checking

* Apply suggestions from code review

Rename tensorboard option

Co-authored-by: Ilya Trushkin <76161256+itrushkin@users.noreply.github.com>

* Remove log level checking

* Pass write_metric function through plan

* Pass write metric function also in workspace

If we dont define this callback in notebook it will be taken from
workspace, if exists

* Add docs for log metric callback

* Clear notebook output

* Add cfssl_scheme.md

* Add diagram as png

* Add description of CA to clients tls connestion

* Fix cfssl scheme

* Set 'metric' tag to train task result in interactive api

It's necessary for logger to parse this tag

* Log metric task result on aggregator additionally.

Collaborator logs metric task result, but sometimes collaborator
and aggregator have different consoles (in interactive api e.g.)

* fix flake8

* Remove default tensorboard option in defaults plan.yaml

* Remove tensorboardX dependency

* Fix some pep8 issues. Remove debugging grpc code.

* VOC shard descriptor

* Remove debugging grpc code.

* First draft working pki with step-ca

* Initial test experiments for new python API

* Director - envoy initial commit

* Base shard descriptor

* Pass shard descriptor to get_collaborator and addition API

* Reorganize project structure for director and collaborator manager

* Move ca commands to separate file

* first LL adaption approach

* fix flake8

* First working example

* Prepare an aggregator workspace and pass initial tensors

* fix flake8

* Adapted launching aggregator on director side

* requirements fix

* Updated the Kvasir notebook

* Add tensorboard to torch_cnn_mnist requirements

* Fix typo

* Initial test experiments for new python API

* Director - envoy initial commit

* Base shard descriptor

* Pass shard descriptor to get_collaborator and addition API

* Reorganize project structure for director and collaborator manager

* VOC shard descriptor

* first LL adaption approach

* Prepare an aggregator workspace and pass initial tensors

* First working example

* CLI commands

* Add tensorboard

* Flake8 and logging

* Flake8

* Setup logging for experiment

* Setup logging for experiment

* Change requirements for grpcio

* Change requirements for grpcio

* Change requirements for grpcio

* Remove unused kwargs

* Resolve aggregator serving

* Resolve aggregator serving

* flake8

* Working notebook

* Create Updated_Kvasir_with_Director.ipynb

* lost fixes after merge

* fixed notebook experimnet start

* Infrastructure for model retrieving

* fixes to the director example

* added requirements file for envoy

* fixed metric logging

* pkg-resources bug fix

* pkg-resources fix

* pip fix

* first try to run and stop aggregator async

* added workspace cleaning in director and envoy

* removing workspace archive on Frontend API
in case of set_new_experiment fail

* Update pki scheme to step-ca flow

* Draft ca integration

* Add ca to utils

* Move ca funtions to separate component

* Add certificates for api layer director and collaborator-manager

* Pass disable_tls option to all grpc clients and servers

* New CLI and fixes for the step-ca component

* flake8 fixes

* Pass certificates name to collaborator and aggregator

* Fix rank_worldsize value

* Fix flake8

* flake8

* Separate install and run command for pki

* Stream metrics to frontend RPC

* UnpackWorkspace context manager

* local changes

* Add step-ca pki mermaid scheme to documentation.

* flake8

* Resolve merge conflicts

* Backward compatibility

* Fix cert api

* make cert paths absolute

* Fix simultaneous requirements installation

* Change default and max token and certificate durations

* Import order

* LogNormal split implementation

* Create DataSplitter interface

* Avoid storing on-disk dataset for lognormal split

* Add random splitter for PyTorch datasets

* Fix linter

* Add dirichlet splits

* Fix linter

* First implementation of API registry

* Added metric logging on frontend

* Tensorboard logs: removed hardcoded name

* Add description of pki workflow to documentation

* Health check and flake8

* Fix pb2 import

* Add info about token live time and other.

* Implement split of shard descriptor

* Fix flake8

* Use numpy arrays in shard descriptor

* Use indices instead of data slices

* Merge fix

* Fix data_splitters package path

* Fix linter

* Update Kvasir shard descriptor

* Fix linter

* Clear notebook

* Add separate splitter for validation data

* Remove PyTorch splitters

* Remove PyTorch dataset from KvasirShardDescriptor

* Change numpy splitter signature

* Fix linter

* Fix Python native tests

* Move Data Splitter functionality to separate shard_descriptor file

* Restore interface signature

* Extract Data Splitter processing in federated_data.py

* Change hash validation in Kvasir Shard Descriptor

* Add docs

* Fix linter

* Fix native tests

Co-authored-by: igor-davidyuk <76463150+igor-davidyuk@users.noreply.github.com>
Co-authored-by: Radionov, Maksim <maksim.radionov@intel.com>
Co-authored-by: maradionov <77286045+maradionov@users.noreply.github.com>
Co-authored-by: Igor Davidyuk <igor.davidyuk@intel.com>
Co-authored-by: amokrov <aleksandr.mokrov@intel.com>
Co-authored-by: aleksandr-mokrov <76171391+aleksandr-mokrov@users.noreply.github.com>

maradionov committed:
Use epochs instead  batch_num. Log current epoch number (#95)

* Use epochs instead  batch_num. Log epoch count

* Fix flake8

* Resolve conversations

* Fix linter

Co-authored-by: Ilya Trushkin <ilya.trushkin@intel.com>

katerina-merkulova committed:
Merge branch 'develop' of https://github.com/katerina-merkulova/openfl into develop

katerina-merkulova committed:
fix requirements

katerina-merkulova committed:
remove repeating downloading

katerina-merkulova committed:
fix downloading logic

katerina-merkulova committed:
fix lint

alexey-gruzdev commented:
@katerina-merkulova thanks for fixes, LGTM!
